Upgrade Pagefind to 1.3.0 and configure Pagefind logging levels

delucis opened the pull request:
<!-- Thank you for opening a PR! We really appreciate you taking the time to help out 🙌 -->

#### Description

- This PR updates Starlight’s Pagefind dependency to 1.3.0.
- Thanks to https://github.com/CloudCannon/pagefind/issues/745 we can now forward Astro’s logging level to Pagefind to reduce (or increase) log verbosity in line with user preference.
- This PR tweaks Pagefind’s default ranking options to preserve result quality, based on testing with Starlight’s own docs site. We ended up going with the [config developed for git-scm.com](https://github.com/git/git-scm.com/blob/c48ff9193a122aad4aad4957298ee7e564e4753a/assets/js/application.js#L342-L349).
- It also exposes Pagefind’s ranking configuration for users who want to tweak it.

##### To-do
- [x] Consider the bundle size changes

astrobot-houston commented:
## size-limit report 📦 
| Path          | Size                  |
| ------------- | --------------------- |
| /index.html   | 7 KB (+0.03% 🔺)      |
| /_astro/*.js  | 25.71 KB (+21.31% 🔺) |
| /_astro/*.css | 13.75 KB (0%)         |

delucis commented:
I’ve repeated @HiDeoo’s exercise from #1750 of comparing results for a selection of search terms in the Starlight docs.

For results that changed rank, I’ve added the difference in ranking compared to Pagefind v1 to help highlight what changed. Bolded entries are results that I’ve subjectively decided are “good” or “best” results, to help track those across columns. I’ve also added a subjective “impact” assessment for each query to summarize my impressions.

### Query: `setup`

**Impact:** 👎 Negative — slightly worse than before, although not terribly

| Rank | 1.0                   | 1.3, termFrequency: `1` | 1.3, termFrequency: `0.5` | 1.3, termFrequency: `0`    |
| :--: | --------------------- | ----------------------- | ------------------------- | -------------------------- |
|  1   | **Manual Setup**      | `+1` Plugins Reference  | `+1` Plugins Reference    | `+1` Plugins Reference     |
|  2   | Plugins Reference     | `-1` **Manual Setup**   | `-1` **Manual Setup**     | `-1` **Manual Setup**      |
|  3   | Using components      | Using components        | Using components          | `+2` Customizing Starlight |
|  4   | **Getting Started**   | **Getting Started**     | **Getting Started**       | `-1` Using components      |
|  5   | Customizing Starlight | Customizing Starlight   | Customizing Starlight     | `-1` **Getting Started**   |

### Query: `installation`

**Impact:** Neutral — we don’t have clear content for this query, closest being “Manual Setup”

| Rank | 1.0                           | 1.3, termFrequency: `1`            | 1.3, termFrequency: `0.5`          | 1.3, termFrequency: `0`            |
| :--: | ----------------------------- | ---------------------------------- | ---------------------------------- | ---------------------------------- |
|  1   | Site Search                   | `+1` Authoring Content in Markdown | `+1` Authoring Content in Markdown | `+1` Authoring Content in Markdown |
|  2   | Authoring Content in Markdown | `-1` Site Search                   | `-1` Site Search                   | `+1` Customizing Starlight         |
|  3   | Customizing Starlight         | Customizing Starlight              | Customizing Starlight              | `-2` Site Search                   |
|  4   | CSS & Styling                 | CSS & Styling                      | CSS & Styling                      | CSS & Styling                      |
|  5   | **Manual Setup**              | **Manual Setup**                   | **Manual Setup**                   | `+1` Configuration Reference       |

### Query: `page`

**Impact:** 👎 Negative — wildly different results, which in the worst case removes the “Pages” result entirely

| Rank | 1.0                   | 1.3, termFrequency: `1`           | 1.3, termFrequency: `0.5`         | 1.3, termFrequency: `0`           |
| :--: | --------------------- | --------------------------------- | --------------------------------- | --------------------------------- |
|  1   | **Pages**             | `+15` Internationalization (i18n) | `+15` Internationalization (i18n) | `+5` Overrides Reference          |
|  2   | Site Search           | `+4` Overrides Reference          | `+4` Overrides Reference          | `+6` Frontmatter Reference        |
|  3   | Overriding Components | `-1` Site Search                  | `-1` Site Search                  | `-2` **Pages**                    |
|  4   | Customizing Starlight | `+4` Frontmatter Reference        | `+4` Frontmatter Reference        | `+7` Configuration Reference      |
|  5   | Eco-friendly docs     | `+6` Configuration Reference      | `+6` Configuration Reference      | `+11` Internationalization (i18n) |
|      |                       | (**Pages** ranked №8)             | (**Pages** ranked №8)             |

### Query: `markdown`

**Impact:** 👎 Negative — downranks the obvious result, which is not ideal

| Rank | 1.0                               | 1.3, termFrequency: `1`                | 1.3, termFrequency: `0.5`              | 1.3, termFrequency: `0`           |
| :--: | --------------------------------- | -------------------------------------- | -------------------------------------- | --------------------------------- |
|  1   | **Authoring Content in Markdown** | `+9` Overrides Reference               | `+9` Overrides Reference               | **Authoring Content in Markdown** |
|  2   | Link Cards                        | `-1` **Authoring Content in Markdown** | `-1` **Authoring Content in Markdown** | `+8` Overrides Reference          |
|  3   | Steps                             | `-1` Link Cards                        | `-1` Link Cards                        | `+2` Pages                        |
|  4   | Card Grids                        | `+1` Pages                             | `+1` Pages                             | Card Grids                        |
|  5   | Pages                             | `-1` Card Grids                        | `-1` Card Grids                        | `-4` Link Cards                   |

### Query: `component`

**Impact:** 👍 Positive — surfaces the same key content, plus adds “Pages” which does include page component docs

| Rank | 1.0                       | 1.3, termFrequency: `1`        | 1.3, termFrequency: `0.5`      | 1.3, termFrequency: `0`        |
| :--: | ------------------------- | ------------------------------ | ------------------------------ | ------------------------------ |
|  1   | **Using components**      | **Using components**           | **Using components**           | `+2` Overrides Reference       |
|  2   | **Overriding Components** | `+19` Pages                    | `+19` Pages                    | `-1` **Using components**      |
|  3   | Overrides Reference       | Overrides Reference            | Overrides Reference            | `-1` **Overriding Components** |
|  4   | Eco-friendly docs         | `-2` **Overriding Components** | `-2` **Overriding Components** | `+2` File Tree                 |
|  5   | Code                      | `+1` File Tree                 | `+1` File Tree                 | `+3` Card Grids                |

### Query: `css`

**Impact:** Neutral — no significant changes for the most part

| Rank | 1.0                     | 1.3, termFrequency: `1`    | 1.3, termFrequency: `0.5`  | 1.3, termFrequency: `0`      |
| :--: | ----------------------- | -------------------------- | -------------------------- | ---------------------------- |
|  1   | **CSS & Styling**       | **CSS & Styling**          | **CSS & Styling**          | **CSS & Styling**            |
|  2   | Icons                   | Icons                      | Icons                      | `+2` Customizing Starlight   |
|  3   | Starlight Showcase      | `+1` Customizing Starlight | `+1` Customizing Starlight | `+2` Configuration Reference |
|  4   | Customizing Starlight   | `-1` Starlight Showcase    | `-1` Starlight Showcase    | `-2` Icons                   |
|  5   | Configuration Reference | Configuration Reference    | Configuration Reference    | `-2` Starlight Showcase      |

### Query: `language`

**Impact:** Neutral — over promotes “Overrides Reference”, but otherwise OK

| Rank | 1.0                                 | 1.3, termFrequency: `1`              | 1.3, termFrequency: `0.5`            | 1.3, termFrequency: `0`            |
| :--: | ----------------------------------- | ------------------------------------ | ------------------------------------ | ---------------------------------- |
|  1   | **Internationalization (i18n)**     | `+4` Overrides Reference             | `+4` Overrides Reference             | **Internationalization (i18n)**    |
|  2   | Link Cards                          | `-1` **Internationalization (i18n)** | `-1` **Internationalization (i18n)** | `+2` Configuration Reference       |
|  3   | Make your docs shine with Starlight | `-1` Link Cards                      | `+1` Configuration Reference         | `+2` Overrides Reference           |
|  4   | Configuration Reference             | Configuration Reference              | `-2` Link Cards                      | `+3` Authoring Content in Markdown |
|  5   | Overrides Reference                 | `+1` Pages                           | `+1` Pages                           | `+3` Sidebar Navigation            |

### Query: `sidebar`

**Impact:** 👎 Negative — only the `termFrequency: 0` variant returns similarly appropriate results

| Rank | 1.0                     | 1.3, termFrequency: `1`      | 1.3, termFrequency: `0.5`    | 1.3, termFrequency: `0`      |
| :--: | ----------------------- | ---------------------------- | ---------------------------- | ---------------------------- |
|  1   | **Sidebar Navigation**  | `+4` Configuration Reference | `+4` Configuration Reference | **Sidebar Navigation**       |
|  2   | Overrides Reference     | Overrides Reference          | Overrides Reference          | Overrides Reference          |
|  3   | Pages                   | Pages                        | Pages                        | `+1` Frontmatter Reference   |
|  4   | Frontmatter Reference   | `-3` **Sidebar Navigation**  | `-3` **Sidebar Navigation**  | `+1` Configuration Reference |
|  5   | Configuration Reference | `-1` Frontmatter Reference   | `-1` Frontmatter Reference   | `-2` Pages                   |

### Query: `lastUpdated`

<details>
<summary><strong>Impact:</strong> Neutral — no change</summary>

| Rank | 1.0                     | 1.3, termFrequency: `1` | 1.3, termFrequency: `0.5` | 1.3, termFrequency: `0` |
| :--: | ----------------------- | ----------------------- | ------------------------- | ----------------------- |
|  1   | Overrides Reference     | Overrides Reference     | Overrides Reference       | Overrides Reference     |
|  2   | Frontmatter Reference   | Frontmatter Reference   | Frontmatter Reference     | Frontmatter Reference   |
|  3   | Configuration Reference | Configuration Reference | Configuration Reference   | Configuration Reference |

</details>

### Query: `plugin`

**Impact:** Neutral — no significant change (reference page second would be nice, but hard to expect the search index to distinguish between guides and reference without assistance)

| Rank | 1.0                          | 1.1                               | 1.1 with `0.5` term frequency     | 1.1 with `0` term frequency       |
| :--: | ---------------------------- | --------------------------------- | --------------------------------- | --------------------------------- |
|  1   | **Plugins and Integrations** | `+1` **Plugins Reference**        | `+1` **Plugins Reference**        | `+1` **Plugins Reference**        |
|  2   | **Plugins Reference**        | `-1` **Plugins and Integrations** | `-1` **Plugins and Integrations** | `-1` **Plugins and Integrations** |
|  3   | Configuration Reference      | CSS & Styling                     | CSS & Styling                     | Configuration Reference           |
|  4   | CSS & Styling                | `+1` Site Search                  | `+1` Site Search                  | CSS & Styling                     |
|  4   | Site Search                  | `-1` Configuration Reference      | `-1` Configuration Reference      | Site Search                       |

astrobot-houston commented:
<!-- lunaria-action-comment -->

## Lunaria Status Overview


🌕 **This pull request will trigger status changes.**

<details>
<summary>Learn more</summary>

By default, every PR changing files present in the [Lunaria configuration's `files` property](https://lunaria.dev/reference/configuration/#files-required) will be considered and trigger status changes accordingly.

You can change this by adding one of the keywords present in the [`ignoreKeywords` property in your Lunaria configuration file](https://lunaria.dev/reference/configuration/#ignorekeywords) in the PR's title (ignoring all files) or by [including a tracker directive](https://lunaria.dev/guides/tracking/#tracker-directives) in the merged commit's description. 
</details>

### Tracked Files




| Locale | File                                                                                                                                                                            | Note                                                       |
| ------ | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ---------------------------------------------------------- |
| en     | [reference/configuration.mdx](https://github.com/withastro/starlight/blob/7b40d95ba3bbf4ff32b3f4ef858d09a130c4eb77/docs%2Fsrc%2Fcontent%2Fdocs%2Freference%2Fconfiguration.mdx) | Source changed, localizations will be marked as outdated.  |

<details>
<summary>Warnings reference</summary>

| Icon | Description                                                                                                                                       |
| ---- | ------------------------------------------------------------------------------------------------------------------------------------------------- |
| 🔄️  | The source for this localization has been updated since the creation of this pull request, make sure all changes in the source have been applied. |

</details>

HiDeoo reviewed (commented):
This is looking great :star2:

Did a diff of the JS files between prod and this PR to see what changed exactly, and most large additions are related to the ability to configure ranking so directly in relation to the PR and also for highlighting support through URL parameters.

Some thoughts:

- As I already had to do that in the past and had to resort to an override re-implementing the whole thing, I could see a world where users would like to customize [`highlightParam`](https://pagefind.app/docs/highlighting/). This would enable users to implement their own highlighting or use the provided script for convenience. Here is what it looks like with this parameter configured and the provided highlighting script where I search for `component` and clicked a result:

   
   <img width="744" alt="image" src="https://github.com/user-attachments/assets/421ef469-a5bd-4344-8959-f2bf5657f899" />

   Definitely not a blocker, and could even be a follow-up PR but wanted to mention it.

- I guess with the logging changes, users may see a lot more warning, e.g. when building the Starlight Docs:

   ```
   Note: Pagefind doesn't support stemming for the language ko.
   Search will still work, but will not match across root words.
   Note: Pagefind doesn't support stemming for the language uk.
   Search will still work, but will not match across root words.
   Note: Pagefind doesn't support stemming for the language ja.
   Search will still work, but will not match across root words.
   Note: Pagefind doesn't support stemming for the language zh-cn.
   Search will still work, but will not match across root words.
   Note: Pagefind doesn't support stemming for the language ko.
   Search will still work, but will not match across root words.
   Note: Pagefind doesn't support stemming for the language uk.
   Search will still work, but will not match across root words.
   Note: Pagefind doesn't support stemming for the language ja.
   Search will still work, but will not match across root words.
   Note: Pagefind doesn't support stemming for the language zh-cn.
   Search will still work, but will not match across root words.
   ```

   Should we consider adding a mention in the docs about thi somehow to mention not all features are supported for all languages?

delucis commented on docs/src/content/docs/reference/configuration.mdx:
```diff
@@ -447,16 +447,32 @@ When using custom themes and setting this to `true`, you must provide at least o
 **default:** `true`
 
-Define whether Starlight’s default site search provider [Pagefind](https://pagefind.app/) is enabled.
+Configure’s Starlight’s default site search provider [Pagefind](https://pagefind.app/).
```
No 😅

delucis commented:
> most large additions are related to the ability to configure ranking so directly in relation to the PR and also for highlighting support through URL parameters

I also just finished looking at the Pagefind UI changes in versions since 1.0.3 (the version currently in use in the Starlight monorepo) to see what is included in the ~4.5 KB (gzipped) bundle size increase.

The main component of this increase seemed me to be the expansion of language support as far as I could tell: Māori, Croatian, Hungarian, Bengali, Vietnamese, Polish, and Danish support was added in v1.0.4; Ukrainian, Romanian, Czech, and Korean added in v1.1.0; Swahili in v1.1.1; and Arabic, Farsi, and Hebrew in v1.2.0. 15 languages with unique strings that won’t compress very well. Kind of a shame in our context where we would have the ability in theory to only ship the current language’s strings, but definitely not worth blocking an update over.

> with the logging changes, users may see a lot more warning

Ah right. This is not strictly speaking our change, it’s just the change in Pagefind 1.3.0 — that warning is [logged with a `v_warn()` method](https://github.com/CloudCannon/pagefind/blame/2a0aa90cfb78bb8551645ac9127a1cd49cf54add/pagefind/src/output/mod.rs#L211) which was [changed from `LogLevel::Verbose` to `LogLevel::Quiet`](https://github.com/CloudCannon/pagefind/pull/759/files#diff-87952b73f108b1c5da7342caeb463393111b0bb3670887a280652b78a1b402aeR107) in 1.3.0.

I think maybe it’s OK on balance and not too likely to impact people as docs in many languages are still reasonably rare, so you probably only hit one of these if any? Not quite sure why the logging is doubled though? Would be good to only log that once per language.

> I could see a world where users would like to customize [`highlightParam`](https://pagefind.app/docs/highlighting/)

Do you think it might make sense to make this a default behaviour with built-in styling in Starlight? I don’t see many drawbacks and seems like a helpful thing to just enable by default and then see if people need to disable it? But then that would probably be best done in a separate PR.

HiDeoo commented:
> The main component of this increase seemed me to be the expansion of language support as far as I could tell

Interesting, I mainly focused on the changes in JS loaded by the browser by diffing loaded code.

I assumed language support would not have a significant impact on this considering that the `pf_meta` file and WASM file are localized, e.g. by visiting the English version the browser loads the `wasm.en.pagefind` & `pagefind.en_4391141799.pf_meta` files while visiting the French version the browser loads the `wasm.fr.pagefind` & `pagefind.fr_9e8fe13141.pf_meta` files.

> Do you think it might make sense to make this a default behaviour with built-in styling in Starlight?

I guess a potential drawback would be URL pollution that some users might not like. Definitely in a follow-up PR as if we decide to go with this by default, I guess we may want to do some research first.

delucis commented:
> I mainly focused on the changes in JS loaded by the browser by diffing loaded code.

Ah yeah, me too, but I only looked at the core UI js file, forgot about other stuff. AFAICT all language strings for the default UI (like “Search”, “Load more...”, “5 results” etc.) are included in that bundle. Although now you mention it, it would be an interesting option if Pagefind could load those dynamically for the current language like they do for other data.

delucis commented on docs/src/content/docs/reference/configuration.mdx:
```diff
@@ -447,16 +447,32 @@ When using custom themes and setting this to `true`, you must provide at least o
 **default:** `true`
 
-Define whether Starlight’s default site search provider [Pagefind](https://pagefind.app/) is enabled.
+Configures Starlight’s default site search provider [Pagefind](https://pagefind.app/).
```
Now I re-read this, actually this should probably be just “Configure” to match the other verbs.
```suggestion
Configure Starlight’s default site search provider [Pagefind](https://pagefind.app/).
```

HiDeoo reviewed (approved):
After the follow-up discussion and potential ideas for future improvements (both on the Starlight side and maybe on the Pagefind side (e.g. UI string)), I'm personally happy with the PR.

Amazing work and glad to finally see this shipping after all the effort put into the research and making sure the default settings are good :tada: :clap: :rocket:
